debacl/utils.py

```diff
@@ -109,6 +113,19 @@ def knn_graph(X, k, method='brute_force', leaf_size=30):
             raise ImportError("The scikit-learn library could not be loaded." +
                               " It is required for the 'ball-tree' method.")
 
+    if method == 'precomputed':
+        # Pseudo-sort each row of the distance matrix so that the indices of the closest k elements are listed first
+        # The entries [...,:k] are not in sorted order by distance
+        neighbors = _np.argpartition(X, k)[:, 0:k]
+
+        # Get the distance values for the first k elements, and then sort 'neighbors' by those distance values
+        radii = X[_np.arange(n)[:, None], neighbors[:]]
```

Comment:
If I'm reading the documentation for `np.argpartition` correctly, I don't think we need lines 122-127, because the `k`th-nearest neighbor is guaranteed to be in position `k`. I think the simpler way to do it would be

```
neighbors = _np.argpartition(X, k - 1)[:, :k]   # note the slight difference from line 119
radii = D[_np.arange(n), neighbors[:, -1]]
```